sdk/src/distributed/worker/pool.rs

```diff
@@ -0,0 +1,345 @@
+use std::{
+    collections::{BTreeMap, VecDeque},
+    sync::Arc,
+    time::Duration,
+};
+
+use sp1_core::{
+    air::PublicValues,
+    runtime::Program,
+    stark::{MachineProver, MachineRecord, RiscvAir},
+    utils::{
+        baby_bear_poseidon2::Val,
+        prove_distributed::{self, ProvingKey},
+        BabyBearPoseidon2, SP1CoreOpts,
+    },
+};
+use sp1_core::{
+    runtime::ExecutionRecord,
+    utils::prove_distributed::{
+        Challenger, Checkpoint, Commitments, PartialProofs, ShardsPublicValues,
+    },
+};
+use tokio::sync::RwLock;
+
+use super::{RequestData, WorkerError, WorkerRequest, WorkerResponse, WorkerSocket};
+
+pub struct WorkerPool {
+    workers: BTreeMap<usize, Arc<RwLock<WorkerSocket>>>,
+}
+
+impl WorkerPool {
+    pub async fn new() -> Result<Self, WorkerError> {
+        let workers = Self::spawn_workers().await?;
+
+        Ok(Self { workers })
+    }
+
+    pub async fn commit<P: MachineProver<BabyBearPoseidon2, RiscvAir<Val>>>(
+        &mut self,
+        prover: &P,
+        program: &Program,
+        elf: &[u8],
+        checkpoints: Vec<(Checkpoint, usize)>,
+        last_checkpoint: Checkpoint,
+        total_nb_shards: u32,
+        public_values: PublicValues<u32, u32>,
+        opts: SP1CoreOpts,
+    ) -> Result<(Commitments, Vec<ShardsPublicValues>), WorkerError> {
+        let nb_checkpoints_per_worker = checkpoints.first().unwrap().1;
+
+        let requests = checkpoints
+            .into_iter()
+            .enumerate()
+            .map(|(i, (checkpoint, nb_checkpoints))| {
+                let state_shard_start =
+                    (i * nb_checkpoints_per_worker * opts.shard_batch_size) as u32;
+
+                WorkerRequest::Commit(RequestData {
+                    elf: elf.to_vec(),
+                    checkpoint,
+                    nb_checkpoints,
+                    public_values,
+                    shard_batch_size: opts.shard_batch_size,
+                    shard_size: opts.shard_size,
+                    state_shard_start,
+                })
+            })
+            .collect();
+
+        let commitments_response = self.distribute_work(requests).await?;
+
+        let mut commitments_vec = Vec::new();
+        let mut shards_public_values_vec = Vec::new();
+        let mut deferred = ExecutionRecord::new(program.clone().into());
+
+        for response in commitments_response {
+            let WorkerResponse::Commitment {
+                commitments,
+                shards_public_values,
+                deferred: mut deferred_new,
+            } = response
+            else {
+                return Err(WorkerError::InvalidResponse);
+            };
+
+            commitments_vec.extend(commitments);
+            shards_public_values_vec.extend(shards_public_values);
+            deferred.append(&mut deferred_new);
+        }
+
+        let mut state = public_values.reset();
+        state.shard = total_nb_shards + 1;
```

Comment:
Remove the +1 here and restore the `state.shard += 1` in deferred checkpoints processing